### PR TITLE
Complement on multi-sets and sub-set iterators

### PR DESCRIPTION
**Pull request by StevellM**
Related to #1238 

This is built on what was already implemented by @fieker : 
- Fix some non-working functions
- Add more documentation and doctests
- Create an html documentation
- Test the different functions of the file
- Add some constructors and functionalities

Not all possible functionalities are available, there are still many existing functions for `Set` and `Dict` which could be implemented for `MSet`; this would have to be done on-demand I guess.

I have changed the printing, to be closer to the printing of `Set` and `Dict`. The code could have been done nicer if julia offered the possibility to use their printing methods for `Dict` where one can choose the separator they want... Instead I wrote something which works similarly (if we see that we will get beyond the size of the screen, we cut and put some dots).

Nothing is perfect or fixed, those are my suggestions of improvements together with a template for the html documentation so that Hecke/Oscar users are aware of the existence of the functionalities for multi-sets.

### CODE REVIEW

**Comment by thofma** [comment]
I have a small question: Could you add the "sum" while you are at it? (see https://en.wikipedia.org/wiki/Multiset#Basic_properties_and_operations).

I will talk to some people about the printing (whether we want to keep `:` or whether we really want to drop the multiplicity if it is one).

**Comment by StevellM** [comment]
Now I noticed that I had to change the `sum(::MSet)` which was falling back into `sun(::AbstractSet)`. Should I add then a deprecation since now we have a proper method `sum(::MSet, ::MSet)` ? Or maybe define `sum(::MSet)` as it was working before ? The latter would imply two different behaviours depending on the number arguments though.

File: src/Misc/MSet.jl [hunk 1]
```diff
@@ -44,70 +191,372 @@ function Base.show(io::IO, ::MIME"text/plain", s::MSet)
...
+
+### Iteration
+
 Base.isempty(s::MSet) = isempty(s.dict)
 Base.length(s::MSet) = sum(values(s.dict))
 Base.IteratorSize(::Type{MSet}) = Base.HasLength()
 Base.IteratorEltype(::Type{MSet}) = Base.HasEltype()
 Base.eltype(::Type{MSet{T}}) where {T} = T
-Base.in(x, s::MSet) = haskey(s.dict, x)
+Base.in(x, s::MSet) = any(y -> x == y, keys(s.dict))
```
Comment (fieker) [inline src/Misc/MSet.jl:263]: Isn't this much slower than the old version? O(n) Vs O(1)?

File: src/Misc/MSet.jl [hunk 2]
```diff
@@ -119,48 +568,92 @@ function Base.filter(pred, s::MSet)
   return t
 end
 
+function Base.filter!(pred, s::MSet)
+  un = unique(s)
+  for x in un
```
Comment (fieker) [inline src/Misc/MSet.jl:573]: This copies everything

File: src/Misc/MSet.jl [hunk 3]
```diff
@@ -44,70 +191,372 @@ function Base.show(io::IO, ::MIME"text/plain", s::MSet)
...
+
+function Base.intersect(s1::MSet, s2::MSet)
+  un1 = unique(s1)
+  un2 = unique(s2)
+  val = filter(x -> any(y -> x == y, un2), un1)
+  T = promote_type(eltype(s1), typeof.(val)...)
+  s = similar(s1, T)
+  d = s.dict
+  for x in val
+    y = un2[findfirst(y -> x == y, un2)]
```
Comment (fieker) [inline src/Misc/MSet.jl:528]: Use isequal(x)?

File: src/Misc/MSet.jl [hunk 4]
```diff
@@ -1,41 +1,188 @@
...
+function multiset(l::Vector{T}, m::Vector{Int}) where {T}
+  @req length(m) == length(l) "Lists must have the same length"
+  @req minimum(m) > 0 "Multiplicity list must consist of positive integers"
+  return MSet{T}(l, m)
+end
+
+@doc raw"""
+    multiset(T::Type) -> MSet{T}
+
+Create an unitialized multi-set `M` with elements of type `T`.
```
Comment (fieker) [inline src/Misc/MSet.jl:125]: It's empty, not uninitialised

File: src/Misc/MSet.jl [hunk 5]
```diff
@@ -44,70 +191,372 @@ function Base.show(io::IO, ::MIME"text/plain", s::MSet)
...
+  un1 = unique(s1)
+  un2 = unique(s2)
+  val = filter(x -> any(y -> x == y, un2), un1)
+  @req promote_type(T, typeof.(val)...) == T "Cannot coerce elements"
+  d = s1.dict
+  for x in un1
+    if !(x in val)
+      delete!(s1, x)
+    else
+      y = un2[findfirst(y -> x == y, un2)]
```
Comment (fieker) [inline src/Misc/MSet.jl:549]: Isequal

File: src/Misc/MSet.jl [hunk 6]
```diff
@@ -119,48 +568,92 @@ function Base.filter(pred, s::MSet)
...
+
+julia> multiplicity(m, 6)
+0
+```
 """
-function multiplicity(s::MSet{T}, x::T) where {T}
-  y = x isa T ? x : T(x)
-  if haskey(s.dict, y)
-    return s.dict[y]
+function multiplicity(s::MSet, x)
```
Comment (fieker) [inline src/Misc/MSet.jl:640]: Change from constant time to linear? Why?

File: src/NumField/NfAbs/Elem.jl [hunk 7]
```diff
@@ -462,7 +462,7 @@ function _ds(fa)
   @assert all(x->x == 1, values(fa.fac))
   T = Int[degree(x) for x = keys(fa.fac)]
   M = MSet(T)
-  return Set(sum(s) for s = subsets(M) if length(s) > 0)
+  return Set(sum(collect(s)) for s = subsets(M) if length(s) > 0)
```
Comment (fieker) [inline src/NumField/NfAbs/Elem.jl:465]: Why collect?

**Review by fieker — CHANGES_REQUESTED** [review]

**Comment by fieker** [comment]
I find the name terrible. Sum(mset) I'd expect a sum over the elements with
multiplicity. Sum of 2 msets I've never seen, or expected

File: src/Misc/MSet.jl (hunk 1 above)
Comment (StevellM) [reply src/Misc/MSet.jl:263]: I haven't tried but the previous version is not practical. See for instance what could happen in a very basic example: 
```
julia> d = Dict{QQFieldElem, Int}(2 => 2)
Dict{QQFieldElem, Int64} with 1 entry:
  2 => 2

julia> haskey(d, 2)
false
```
I have not thought about about a better work-around..

File: src/Misc/MSet.jl (hunk 6 above)
Comment (StevellM) [reply src/Misc/MSet.jl:640]: Same as explain for the `in` function: if I have a `MSet{QQFieldElem}`, then for testing the multiplicity of 2, one needs to check for `QQ(2)`. If we want to force this and then do not leave room for this kind of flexibility, I am ok with reverting this.

**Comment by StevellM** [comment]
The naming `sum` for `MSet` is taken from what I have seen on Wikipedia. I had a look to other CAS: python apparently uses "combine" and maple uses the "+" operator. The thing with `sum(::MSet{T})` is that it would only work whenever `T <: RingElem` or at least the types for which `sum` is defined. We could change `sum` as I defined it with something of the form `multiset_sum`, `disjoint_union` or `combine` and keep maybe the `+` operator for simplicity ? In that case the `sum(::MSet)` will use directly `sum(::AbstractSet)` as it was doing before (provided that the user know what they are doing)

**Comment by fieker** [comment]
On Tue, Oct 10, 2023 at 11:29:17PM -0700, Stevell Muller wrote:
> @StevellM commented on this pull request.
> 
> 
> 
> >  Base.isempty(s::MSet) = isempty(s.dict)
>  Base.length(s::MSet) = sum(values(s.dict))
>  Base.IteratorSize(::Type{MSet}) = Base.HasLength()
>  Base.IteratorEltype(::Type{MSet}) = Base.HasEltype()
>  Base.eltype(::Type{MSet{T}}) where {T} = T
> -Base.in(x, s::MSet) = haskey(s.dict, x)
> +Base.in(x, s::MSet) = any(y -> x == y, keys(s.dict))
> 
> I haven't tried but the previous version is not practical. See for instance what could happen in a very basic example: 
> ```
> julia> d = Dict{QQFieldElem, Int}(2 => 2)
> Dict{QQFieldElem, Int64} with 1 entry:
>   2 => 2
> 
> julia> haskey(d, 2)
> false
> ```
> I have not thought about about a better work-around..
> 
That was a deliberate decision to not support that. I'm (un)happy to
discuss this in detail:
 In Julia base, effort is made so that 
   hash(1) == hash(1//1) == hash(1.0) == ...
Despite them being different types. This is possible, for BigInt and
BigFloat only via some memory expensive operations, the float is
decomposed into a rational and then mutilated.
This screws up the performance in many important algorithms.
Thus we (I) decided to not support that madness
  hash functions should never allocate
The price is that for ZZRingElem(2)^100 the hash is different to
Int128(2)^100...
For in: it would be better, in my opinion to coerce into the type if we
want to support that. But the entire point of Dicts is to have an O(1)
lookup and not an O(length(D))
in(a::T, M::MSet{T}) = haskey(M.dict, a)
in(a::Any, M::MSet{T}) = haskey(M.dict, T(a))
it will still be "bad" if you put e.g. number field elements in - esp. 
in different fields....

**Comment by fieker** [comment]
On Tue, Oct 10, 2023 at 11:43:44PM -0700, Stevell Muller wrote:
> The naming `sum` for `MSet` is taken from what I have seen on Wikipedia. I had a look to other CAS: python apparently uses "combine" and maple uses the "+" operator. The thing with `sum(::MSet{T})` is that it would only work whenever `T <: RingElem` or at least the types for which `sum` is defined. We could change `sum` as I defined it with something of the form `multiset_sum`, `disjoint_union` or `combine` and keep maybe the `+` operator for simplicity ? In that case the `sum(::MSet)` will use directly `sum(::AbstractSet)` as it was doing before (provided that the user know what they are doing)
> 
In the end I don't mind. I've never tried to add MSets and don't see
this happening in the near future....
sum and prod for Sets I haven't tried recently. I think it was not
allowed for a long time....

**Comment by StevellM** [comment]
Thanks for the explanation! I will revert what I have changed to keep it in linear time then. I have modified also some other parts to avoid to create unnecessary lists.

For the sum, I will just keep the `+` for now and if one day one comes up with an acceptable name, it will be easy to add.

**Comment by StevellM** [comment]
@fieker Do the last changes I have been seem good for you ?

**Review by fieker — APPROVED** [review]

**Comment by thofma** [comment]
Merci Stevell

**Pull request merged**